Memory corruption fixes (found by Microsys while porting to LS1046)

Kay-Potthoff opened the pull request:
Compiled and ported sja1105-tool on LS1046. The fixes are necessary, because the tool terminated
with segmentation fault or occasionally with 'unmap: invalid pointer'.

Commit:
added .gitignore

Signed-off-by: Kay Potthoff <Kay.Potthoff@microsys.de>

Commit:
common.h: added robust FREE macro

The use of the macro FREE() should replace all calls of free() for
achieving higher robustness.

Signed-off-by: Kay Potthoff <Kay.Potthoff@microsys.de>

Commit:
staging-area.c: closing file handle before leaving staging_area_save()

In case that reliable_write() returns with an error the file handle
'fd' should be closed before leaving the function.

Signed-off-by: Kay Potthoff <Kay.Potthoff@microsys.de>

Commit:
strings.c: fixed deallocation of cmd_matches[]

In case of an absolute match all strings allocated with strdup()
had not been deallocated so far. Now the memory is correctly freed
when the routine is left via 'goto out'.

Signed-off-by: Kay Potthoff <Kay.Potthoff@microsys.de>

Commit:
parse-status.c: memory not freed on error

When sja1105_port_status_get() returns with an error the memory
previously allocated and pointed by print_buf[] needs to be freed.

Signed-off-by: Kay Potthoff <Kay.Potthoff@microsys.de>

Commit:
config-show.c: fixed segmentation fault

The tool terminated with a segmentation fault, because it did a free()
on an uninitialized pointer. This is fixed via initializing the pointers
with NULL. Additionally, the code now avoids freeing NULL pointers.

Signed-off-by: Kay Potthoff <Kay.Potthoff@microsys.de>

Commit:
removed const attribute in struct sja1105_spi_setup

The attribute const has been removed for the fields device and
staging_area because these two fields can point to dynamically
allocated memory, too. This memory then needs to be freed leading to
the situation that a const pointer will be freed. Now the memory is always
allocated dynamically and thus always correctly freed. This increases
robustness and portability to other platforms, e.g. when running on LS1046.

Signed-off-by: Kay Potthoff <Kay.Potthoff@microsys.de>

vladimiroltean reviewed (changes requested):
I don't think this is really correct. You shouldn't abuse the FREE macro by deliberately leaving two free() loops in the same code path.

vladimiroltean reviewed (changes requested):
Indentation, spacing.

Commit:
parse-status.c: fixed memory error in status_ports()

When sja1105_port_status_get() returns with an error the function was
left without freeing the memory pointed by print_buf[].

Signed-off-by: Kay Potthoff <Kay.Potthoff@microsys.de>

Commit:
config-show.c: replaced FREE() macro

The use of the macro FREE() has been replaced with an explicit
call of the system function free().

Signed-off-by: Kay Potthoff <Kay.Potthoff@microsys.de>

Commit:
staging-area.c: fixed memory errors

When the function sja1105_static_config_hexdump() or the function
sja1105_static_config_unpack() returns with an error the memory
pointed by 'buf' and the file handle 'fd' are now correctly freed.

Signed-off-by: Kay Potthoff <Kay.Potthoff@microsys.de>

Commit:
tool-config-file.c: replaced FREE() macro

The use of the macro FREE() has been replaced with an explicit
call of the system function free().

Signed-off-by: Kay Potthoff <Kay.Potthoff@microsys.de>

Commit:
main.c: replaced FREE() macro

The use of the macro FREE() has been replaced with an explicit
call of the system function free().

Signed-off-by: Kay Potthoff <Kay.Potthoff@microsys.de>

Commit:
strings.c: added checks for memory allocation in linewise_concat()

The pointers 'next_line' and 'current_line' are now checked if
calloc() succeeded. This make obsoletes the use of the FREE()
macro when leaving the function.

Signed-off-by: Kay Potthoff <Kay.Potthoff@microsys.de>

Commit:
common.h: removed FREE() macro

This macro is no longer used in the code.

Signed-off-by: Kay Potthoff <Kay.Potthoff@microsys.de>

Kay-Potthoff commented:
Hi,
as you suggested I have removed the macro FREE() and replaced it with explicit calls of free().
Additionally, I fixed the problems that in some error cases allocated resources have not been freed
correctly.
-Kay

vladimiroltean commented:
Do you mind if I make changes to your patchset?
There are still things that need to change:
* Squash the addition of the FREE() macro with its removal.
* The `staging-area.c: fixed memory errors` patch is duplicating some cleanup code instead of fixing the error codepath. If you're fixing it you might as well do it properly.
* Patch `parse-status.c: fixed memory error in status_ports()` is still using calloc memory without NULL-checking it.
* Some NULL checks are mindless e.g. `main.c: replaced FREE() macro` where you end up checking `if (spi_setup->staging_area)` twice. There are other places as well.
